core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -694,7 +694,7 @@ private[storage] class PartiallyUnrolledIterator[T](
   }
 
   override def next(): T = {
-    if (unrolled == null) {
+    if (unrolled == null || unrolled.isEmpty) {
```

Comment:
@wangyum This bug happens because in `TorrentBroadcast.readBroadcastBlock`, we directly call `next()` without calling `hasNext()` first which can null out `unrolled`.

You can check out
https://github.com/apache/spark/blob/7d5cb3af7621ad6eb85d1ba7f585c3921ca0a242/core/src/main/scala/org/apache/spark/broadcast/TorrentBroadcast.scala#L210

Calling `hasNext()` before `next()` can fix this too. But this fixing is simpler and I think it should be ok.